src/Campaigns/Migrations/Donations/AddCampaignId.php

```diff
@@ -0,0 +1,88 @@
+<?php
+
+namespace Give\Campaigns\Migrations\Donations;
+
+use Give\Donations\ValueObjects\DonationMetaKeys;
+use Give\Framework\Database\DB;
+use Give\Framework\Database\Exceptions\DatabaseQueryException;
+use Give\Framework\Migrations\Contracts\Migration;
+use Give\Framework\Migrations\Exceptions\DatabaseMigrationException;
+
+/**
+ * @unreleased
+ */
+class AddCampaignId extends Migration
+{
+    /**
+     * @inheritDoc
+     */
+    public static function id(): string
+    {
+        return 'add_campaign_id_to_donations';
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public static function title(): string
+    {
+        return 'Add campaign id to donations';
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public static function timestamp(): string
+    {
+        return strtotime('2024-11-22 00:00:00');
+    }
+
+    /**
+     * @inheritDoc
+     * @throws DatabaseMigrationException
+     */
+    public function run()
+    {
+        $relationships = [];
+
+        try {
+
+            $data = DB::table('give_campaign_forms')
+                ->select('campaign_id', 'form_id')
+                ->getAll();
+
+            foreach ($data as $relationship) {
+                $relationships[$relationship->campaign_id][] = $relationship->form_id;
+            }
+
+            $donations = DB::table('posts')
+                ->select('ID')
+                ->attachMeta(
+                    'give_donationmeta',
+                    'ID',
+                    'donation_id',
+                    [DonationMetaKeys::FORM_ID(), 'formId']
+                )
+                ->where('post_type', 'give_payment')
+                ->getAll();
+
+            foreach($donations as $donation) {
+                foreach ($relationships as $campaignId => $formIds) {
+                    if (in_array($donation->formId, $formIds)) {
+                        DB::table('give_donationmeta')
+                            ->insert([
+                                'donation_id' => $donation->ID,
+                                'meta_key' => DonationMetaKeys::CAMPAIGN_ID,
+                                'meta_value' => $campaignId,
+                            ]);
```

Comment:
Seems like this could be accomplished as a single insert of many rows?

Reply:
Yes, with SQL. With WPDB, not really. I would love to have that feature. Maybe we should consider updating the `insert` method to support that.

Reply:
Seems like a shame to knowingly have a nasty N+1 issue here that we could definitely circumvent just because of a lack of a WPDB feature. I think it's worth taking the moment to build the insert many feature for this.

Reply:
@JasonTheAdams, I have been thinking about this in the past. Since data for inserting will probably be collected inside the loop, I thought having something like this might be handy: 

```php

        $qb = DB::table('give_donationmeta');

        foreach($data as $row) {
            $qb->set([
                'donation_id' => $row->id,
                'meta_key' => $row->key,
                'meta_value' => $row->value,
            ]);
        }

        $qb->insert();
```

This approach will require altering the `insert` method to be something like this:

```php
    public function insert($data = [], $format = null)
    {
        // collection represents collected data for insert
        if ($this->collection instanceof Collection) {
            $this->insertMany($this->collection);
        }
        
        if (empty($data)) {
            throw new InvalidArgumentException('data cannot be empty');
        }
        
        return DB::insert(
            $this->getTable(),
            $data,
            $format
        );
    }
```